### ISSUE TITLE
Shared array memory corruption when fission bank is full

### ISSUE THREAD

**Issue by paulromano** [openmc-dev/openmc#1853]
I've noticed that on a model where the initial k-effective is far from 1.0 and the fission bank fills up, instead of just showing the normal warning:
```
 WARNING: The shared fission bank is full. Additional fission sites created in
          this generation will not be banked.
```
we actually end up getting a `double free or corruption (!prev)` at the end of `openmc::sort_fission_bank` and the run aborts altogether. So far I haven't been able to figure out why this is -- in the `SharedArray::thread_safe_append` method, it looks like proper protections are in place such that we don't write off the end of the array even when two threads are trying to simultaneously add entries at the end. I should also mention that this error does not happen if only one thread is in use. Here's a model based on the pincell examples that reproduces this behavior:
```Python
from math import log10
import numpy as np
import openmc

uo2 = openmc.Material(name='UO2 fuel at 2.4% wt enrichment')
uo2.set_density('g/cm3', 10.29769)
uo2.add_element('U', 1e-3, enrichment=1e-3)
uo2.add_element('O', 2.)

helium = openmc.Material(name='Helium for gap')
helium.set_density('g/cm3', 0.001598)
helium.add_element('He', 2.4044e-4)

zircaloy = openmc.Material(name='Zircaloy 4')
zircaloy.set_density('g/cm3', 6.55)
zircaloy.add_element('Sn', 0.014  , 'wo')
zircaloy.add_element('Fe', 0.00165, 'wo')
zircaloy.add_element('Cr', 0.001  , 'wo')
zircaloy.add_element('Zr', 0.98335, 'wo')

borated_water = openmc.Material(name='Borated water')
borated_water.set_density('g/cm3', 0.740582)
borated_water.add_element('B', 4.0e-5)
borated_water.add_element('H', 5.0e-2)
borated_water.add_element('O', 2.4e-2)
borated_water.add_s_alpha_beta('c_H_in_H2O')

# Collect the materials together and export to XML
materials = openmc.Materials([uo2, helium, zircaloy, borated_water])
materials.export_to_xml()

# Create cylindrical surfaces
fuel_or = openmc.ZCylinder(r=0.39218, name='Fuel OR')
clad_ir = openmc.ZCylinder(r=0.40005, name='Clad IR')
clad_or = openmc.ZCylinder(r=0.45720, name='Clad OR')

# Create a region represented as the inside of a rectangular prism
pitch = 1.25984
box = openmc.rectangular_prism(pitch, pitch, boundary_type='reflective')

# Create cells, mapping materials to regions
fuel = openmc.Cell(fill=uo2, region=-fuel_or)
gap = openmc.Cell(fill=helium, region=+fuel_or & -clad_ir)
clad = openmc.Cell(fill=zircaloy, region=+clad_ir & -clad_or)
water = openmc.Cell(fill=borated_water, region=+clad_or & box)

# Create a geometry and export to XML
geometry = openmc.Geometry([fuel, gap, clad, water])
geometry.export_to_xml()

# Indicate how many particles to run
settings = openmc.Settings()
settings.batches = 100
settings.inactive = 10
settings.particles = 1000

# Create an initial uniform spatial source distribution over fissionable zones
lower_left = (-pitch/2, -pitch/2, -1)
upper_right = (pitch/2, pitch/2, 1)
uniform_dist = openmc.stats.Box(lower_left, upper_right, only_fissionable=True)
settings.source = openmc.source.Source(space=uniform_dist)
settings.export_to_xml()
```

@jtramm Let me know if you have any ideas here.

**Comment by jtramm** [comment]
Ah interesting. After a quick look through nothing is jumping out at me as being obviously wrong either -- I'll take a closer look and track down the issue.

**Comment by jtramm** [comment]
I think I've tracked down what the problem is. The memory error occurs when sorting the fission bank and is caused by there being invalid offsets into the bank resulting in writing off the end of the array. This is due to how the sorting algorithm works based on the particle & progeny ID and how we are tracking the number of fission sites generated per particle. Basically, when we have too many fission sites and protect the fission bank from overflowing, this causes an unaccounted for mismatch between the size of the fission bank and the sum of the number of fission sites generated by all particles.

Sites are not being pushed back past the end of the fission bank itself, but, particles are still having their `n_progeny_` variable incremented which results in the sorting function going past where it should. Looking back on the commit history, I found that this error was actually not in the original shared fission bank implementation but only appeared in an edit I made in 76e58e7b4c064dda2c6a3ad94dd08f128c5c777f. This commit was in response to a review request to change the SharedArray interface to be more C++ STL-like. Originally, we could call the append function to get an index that we could write to. If the index was -1, we did not prepare a fission site and did not increment the particle `n_progeny_` variable, so the original accounting I think was correct. However, when the change was made to the interface to make it require appending an object rather than returning an index to write to (akin to a C++ push_back operation), this now meant that we did not know until after `n_progeny_` was incremented whether or not the site was actually going to fit in the fission bank, meaning that the sorting operation would be doomed to write off the end of the array later on. This was definitely an error I should have caught when refactoring the interface as part of the integration of PR #1461. My guess is I did not forsee that just refactoring the interface would introduce logic implications like this so I did not go back through check all the corner cases again.

Another thing to note is that I believe that the current code works in serial because the order that the particle sites get added to the fission bank is deterministic, starting from low particle IDs to high and from low progeny numbers to high. This means that the sorting loop stops before needing to write off the end of the array as everything is already in order. However, when run with 2 or more threads, some of the higher parent particle ID's will appear earlier in the fission bank, and when sorting occurs it will try to move them past the end of the array.

There's a few different ways of fixing, but I think the easiest is probably to just decrement the `n_progeny_` variable after a failed append attempt on the fission bank. I want to do some testing though to make sure there are no other corner cases I missed as a result of the changed interface.
